File: VirtualOrigenApp/virtual_origen_app/lib/services/repository/inversor_now/inversor_now_firebase_repository.dart
import 'dart:async';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:virtual_origen_app/models/inversor_now.dart';
import 'package:virtual_origen_app/services/repository/gen_firebase/gen_firebase_repository_only_get.dart';

class InversorNowFirebaseRepository
    extends GenFirebaseRepositoryOnlyGet<InversorNow, String> {
  @override
  String get collectionName => "inversor_now";

  @override
  String get idName => throw UnimplementedError();

  @override
  String get listName => throw UnimplementedError();

  @override
  InversorNow fromJson(Map<String, dynamic> json) {
    return InversorNow(
      battery: json["battery"],
      consumption: json["consumption"],
      gain: json["gain"],
    );
  }

  @override
  void addListener(
      {required String idc, required Function(List<InversorNow>) listener}) {
    listenerStream = FirebaseFirestore.instance
        .collection(collectionName)
        .doc(idc)
        .snapshots()
        .listen((DocumentSnapshot snapshot) {
      if (checkIfDocExists(snapshot)) return;
      var data = snapshot.data() as Map<String, dynamic>;
      var entity = fromJson(data);
      listener([entity]);
    });
  }

  @override
  Future<List<InversorNow>> findAll({required String idc}) async {
    var data = await findById(id: idc, idc: idc);
    return data != null ? [data] : [];
  }

  @override
  Future<InversorNow?> findById(
      {required String id, required String idc}) async {
    var docRef = await getRefAndSnapshot(idc: idc);
    if (checkIfDocExists(docRef.value)) return null;
    var data = docRef.value!.data() as Map<String, dynamic>;
    return fromJson(data);
  }
}

File: VirtualOrigenServer/VirtualOrigenServer/SmartDevicesService/Services/Repository/InversorNow/InversorNowFirebaseRepository.cs
using Google.Cloud.Firestore;
using SmartDevicesService.Models;
using SmartDevicesService.Services.Db;

namespace SmartDevicesService.Services.Repository.InversorNow;

public class InversorNowFirebaseRepository : IInversorNow
{
    private readonly CollectionReference collection;
    private readonly string _collectionName = "inversor_now";

    public InversorNowFirebaseRepository()
    {
        var db = MyFirebaseDb.Instance;
        collection = db.Db.Collection(_collectionName);
    }

    private static InversorNowFirebaseRepository? _instance;

    public static InversorNowFirebaseRepository Instance
    {
        get
        {
            _instance ??= new InversorNowFirebaseRepository();
            return _instance;
        }
    }

    private async Task<DocumentSnapshot?> GetSnapshotAsync(string idc)
    {
        DocumentReference document = collection.Document(idc);
        DocumentSnapshot snapshot = await document.GetSnapshotAsync();
        if (!snapshot.Exists)
        {
            return null;
        }
        return snapshot;
    }

    public async Task<InversorData?> GetByIdcAsync(string idc)
    {
        var snapshot = await GetSnapshotAsync(idc);
        if (snapshot == null)
        {
            return null;
        }
        var data = snapshot.ToDictionary();
        return new InversorData(data);
    }
}
